feature(esp_tinyusb): v2.0.0 README.md update (#191)
espressif/esp-usb: device/esp_tinyusb/README.md

Values of default descriptors could be configured via `menuconfig`.

```c
  #include "tinyusb_default_config.h"

  void main(void) {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.descriptor.device = &custom_device_descriptor;
    tusb_cfg.descriptor.full_speed_config = custom_full_speed_configuration;
#if (TUD_OPT_HIGH_SPEED)
    tusb_cfg.descriptor.high_speed_config = custom_high_speed_configuration;
#endif // TUD_OPT_HIGH_SPEED

    tinyusb_driver_install(&tusb_cfg);
  }
```

### USB PHY configuration & Self-Powered Device

For self-powered devices, monitoring the VBUS voltage is required. To do this:

- Configure a GPIO pin as an input, using an external voltage divider or comparator to detect the VBUS state.
- Set `self_powered = true` and assign the VBUS monitor GPIO in the `tinyusb_config_t` structure.

```c
  #include "tinyusb_default_config.h"

  void main(void)
  {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.phy.self_powered = true;
    tusb_cfg.phy.vbus_monitor_io = GPIO_NUM_0;

    tinyusb_driver_install(&tusb_cfg);
  }
```
If external PHY is used:

```c

>   #include "tinyusb_default_config.h"
> 
>   void main(void)
>   {
>     tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
> 
>     tusb_cfg.phy.skip_setup = true;
> 
>     tinyusb_driver_install(&tusb_cfg);
>   }

Reviewer: Could be useful for external PHY users :)
```suggestion
  #include "tinyusb_default_config.h"
  #include "esp_private/usb_phy.h"

  void main(void)
  {
    // First create the ext. PHY
    usb_new_phy(...);
    
    // Install esp_tinyusb
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
    tusb_cfg.phy.skip_setup = true;
    tinyusb_driver_install(&tusb_cfg);
  }
```

This component adds features to TinyUSB that help users with integrating TinyUSB with their ESP-IDF application.

> Provides default and custom configuration of TinyUSB and allows to run USB devices on ESP chips.

Reviewer 2: ```suggestion
Provides default and custom configuration of TinyUSB and allows to run USB devices on USB-ORG Enabled ESP chips.
```

Or something similar. Not to give an impression, that the device can be run on any esp chip

```
idf.py add-dependency esp_tinyusb~1.0.0
```

## Breaking Changes

> - [v2.0.0]((https://docs.espressif.com/projects/esp-usb/docs/migration-guide/v2)

Author: ```suggestion
- [v2.0.0](https://docs.espressif.com/projects/esp-usb/docs/migration-guide/v2)
```

When several peripheral ports are available by the hardware, the specific port could be configured manually:

```c
  #include "tinyusb_default_config.h"

  void main(void) {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
#if (TUD_OPT_HIGH_SPEED)
    tusb_cfg.port = TINYUSB_PORT_HIGH_SPEED_0;
#else
    tusb_cfg.port = TINYUSB_PORT_FULL_SPEED_0;
#endif
    tinyusb_driver_install(&tusb_cfg);
  }
```

### Task Configuration

When the default parameters of the internal task should be changed:

```c
  #include "tinyusb_default_config.h"

  void main(void) {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
    tusb_cfg.task = TINYUSB_TASK_CUSTOM(4096 /*size */, 4 /* priority */, 0 /* affinity: 0 - CPU0, 1 - CPU1 ... */);
    tinyusb_driver_install(&tusb_cfg);
  }
```

### USB Descriptors Configuration

Configure USB descriptors using the `tinyusb_config_t` structure:

- `descriptor.device`
- `descriptor.string`
- `descriptor.full_speed_config`
- `descriptor.high_speed_config`
- `descriptor.qualifier`

If any descriptor field is set to `NULL`, default descriptor will be assigned during installation.
Values of default descriptors could be configured via `menuconfig`.

> Advanced configuration could be done in advance: to set the descriptors or choose the peripheral port.

Author: ```suggestion

```

```c
  #include "tinyusb_default_config.h"

  void main(void) {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.descriptor.device = &custom_device_descriptor;
    tusb_cfg.descriptor.full_speed_config = custom_full_speed_configuration;
#if (TUD_OPT_HIGH_SPEED)
    tusb_cfg.descriptor.high_speed_config = custom_high_speed_configuration;
#endif // TUD_OPT_HIGH_SPEED

    tinyusb_driver_install(&tusb_cfg);
  }
```

### PHY Configuration & Self-Powered Device

> Self-powered devices must monitor VBUS voltage. Use a GPIO pin with a voltage divider or comparator to detect VBUS state. Set `self_powered = true` and assign the VBUS monitor GPIO in `tinyusb_config_t`:

Author: Link to the comparator section in esp-idf docs.

  .callback_line_coding_changed = NULL
};
tusb_cdc_acm_init(&acm_cfg);
```

Redirect standard I/O streams to USB with `esp_tusb_init_console` and revert with `esp_tusb_deinit_console`.

### USB Mass Storage Device (MSC)

To enable Mass Storage Device:

- select the option from `menuconfig`
- configure storage for MSC Device class: SPI Flash or SD/MMC (when supported by the hardware).

**SPI-Flash Storage**

```c
static esp_err_t storage_init_spiflash(wl_handle_t *wl_handle)
{
  // ... partition and mount logic ...
}

void main(void)
{
  wl_handle_t wl_handle;
  storage_init_spiflash(&wl_handle);

  tinyusb_msc_storage_handle_t storage_hdl;
  const tinyusb_msc_storage_config_t cfg = {
    .medium.wl_handle = wl_handle,
  };
  tinyusb_msc_new_storage_spiflash(&config_spi, &storage_hdl);
}
```

**SD-Card Storage**

```c
static esp_err_t storage_init_sdmmc(sdmmc_card_t **card)
{
  // ... SDMMC host and slot config ...
}

void main(void)
{
  sdmmc_card_t *card;
  storage_init_sdmmc(&card);

  tinyusb_msc_storage_handle_t storage_hdl;
  const tinyusb_msc_storage_config_t cfg = {
    .medium.card = card,
  };
  tinyusb_msc_new_storage_sdmmc(&config_spi, &storage_hdl);
}
```

> **Storage callback**
> Storage event callback is called, when one of the following events occurred:

Author: ```suggestion
**Storage callback**

Storage event callback is called, when one of the following events occurred:
```

```
idf.py add-dependency esp_tinyusb~1.0.0
```

> 

Reviewer 2: The mainifest yaml file, should not have `esp_tinyusb: "~1.0.0"` version 1 anymore. Same for the `idf.py` command

> ## Breaking Changes

Reviewer 2: Here would make sense to add some comments, about breaking changes relation (between which `esp_tinyusb` versions are the breaking changes)

For example here, it  is a migration guide from `esp_tinyusb` version `1.x.x` to version `2.0.0`

> - [v2.0.0](https://docs.espressif.com/projects/esp-usb/docs/migration-guide/v2)

Reviewer 2: I understand, that the docs page for esp-usb is about to be created? Now, it's just a dead link.

## USB Device Stack usage, installation & configuration

> Hardware-related documentation could be found in [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s2/api-reference/peripherals/usb_device.html).

Reviewer 3: Small wording suggestion: “can be found” sounds a bit clearer than “could be found” here.

Overall, LGTM—nice update, and I checked that the links are working!

### Structure overview

The Device Stack is built on top of TinyUSB and provides:

- Custom USB descriptor support
- Serial device (CDC-ACM) support
- Standard stream redirection through the serial device
- Storage media support (SPI-Flash and SD-Card) for USB MSC Class
- A dedicated task for TinyUSB servicing

> ### Build-time configuration options

Reviewer 2: Since we moved many Built-time configs to run-time configs. Could you maybe provide a paragraph, where you would list some of the run-time confgis?

Configure the Device Stack using `menuconfig`:

- TinyUSB log verbosity
- Default Device/String descriptors options
- Class-specific options

> > **⚠️ Important:** For demonstration purposes, all error handling logic is removed from the code. Please do not ignore these details in actual development.

Reviewer 2: Is this comment related to the code snippets below? If so, you could put it bit more below, under the `Installation` paragraph for example. Right now, it looks like it's a part of `Build-time configuration options`

### Installation

Install the Device Stack by calling `tinyusb_driver_install` with a `tinyusb_config_t` structure.
Default configuration is available with `TINYUSB_DEFAULT_CONFIG()` macros.

> Default installation configures the port (High-speed port when available by hardware, Full-speed otherwise), task (with default parameters) and descriptors.

Reviewer 2: You can list all, what is configured within the default config. Up to you.

```suggestion
Default installation configures the port (High-speed port when available by hardware, Full-speed otherwise), task (with default parameters), descriptors, PHY and event callbacks.
```

Default descriptors are available for the following USB classes: CDC, MSC and NCM.

```c

>   #include "tinyusb_default_config.h"

Reviewer 2: I believe it's `tinyusb_default_configs.h` not `tinyusb_default_config.h`. Same comment for all the code snippets provided in this readme.

```suggestion
  #include "tinyusb_default_configs.h"
```

    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();
    tusb_cfg.task = TINYUSB_TASK_CUSTOM(4096 /*size */, 4 /* priority */, 0 /* affinity: 0 - CPU0, 1 - CPU1 ... */);
    tinyusb_driver_install(&tusb_cfg);
  }
```

### USB Descriptors Configuration

Configure USB descriptors using the `tinyusb_config_t` structure:

- `descriptor.device`
- `descriptor.string`
- `descriptor.full_speed_config`
- `descriptor.high_speed_config`
- `descriptor.qualifier`

If any descriptor field is set to `NULL`, default descriptor will be assigned during installation.
Values of default descriptors could be configured via `menuconfig`.

```c
  #include "tinyusb_default_config.h"

  void main(void) {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.descriptor.device = &custom_device_descriptor;
    tusb_cfg.descriptor.full_speed_config = custom_full_speed_configuration;
#if (TUD_OPT_HIGH_SPEED)
    tusb_cfg.descriptor.high_speed_config = custom_high_speed_configuration;
#endif // TUD_OPT_HIGH_SPEED

    tinyusb_driver_install(&tusb_cfg);
  }
```

### PHY Configuration & Self-Powered Device

Self-powered devices must monitor VBUS voltage. To achieve this, implement:

> - Enable GPIO pin with a voltage divider or comparator to detect VBUS state

Reviewer 2: `Configure GPIO` instead of `Enable GPIP`. Up to you.

```suggestion
- Configure GPIO pin as an input, with an external voltage divider or comparator to detect VBUS state
```

- Set `self_powered = true` and assign the VBUS monitor GPIO in `tinyusb_config_t`

```c
  #include "tinyusb_default_config.h"

  void main(void)
  {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.phy.self_powered = true;
    tusb_cfg.phy.vbus_monitor_io = GPIO_NUM_0;

    tinyusb_driver_install(&tusb_cfg);
  }
```
If external PHY is used:

```c
  #include "tinyusb_default_config.h"

  void main(void)
  {
    tinyusb_config_t tusb_cfg = TINYUSB_DEFAULT_CONFIG();

    tusb_cfg.phy.skip_setup = true;

    tinyusb_driver_install(&tusb_cfg);
  }
```

> 

Reviewer 2: Just to put a new 2nd layer paragraph, since you are starting to write about device classes.

```suggestion

## Device classes

```
